src/hooks/useGetPrice.ts

```diff
@@ -1,12 +1,49 @@
 import { useState, useCallback, useEffect } from "react";
 import Decimal from "decimal.js";
+import NodeCache from "node-cache";
 
 import { getOneinchPrice } from "@gnosis.pm/dex-liquidity-provision/scripts/utils/price_utils";
 
 import { TokenDetails } from "types";
+import { ONE_DECIMAL, PRICE_CACHE_TIME } from "utils/constants";
 
 export type PriceSources = "1inch";
 
+const cache = new NodeCache({ stdTTL: PRICE_CACHE_TIME });
+
+/**
+ * Builds 2 cache keys:
+ * 1. Regular for the current token pair
+ * 2. Inverted in case regular is not found.
+ *  We can then invert the price without another query
```

Comment:
In general this is not true. The reason is they are different prices, one is the best ask price, the other the best bid price. So they will differentiate by the spread + slippage that exists on 1inch. For tokens with low liquidity this can be quite significant.

One extreme example to illustrate how these two prices are difference (although not really applicable here, just to serve as an example) is right before a token first goes on sale. Then it has a price in one direction, this comes from all the bids waiting to buy the new token, but an "infinite" price (i.e. cannot be bought) in the opposite order, since the token is not yet available.

That being said, maybe this difference is not so bad for this specific use case, since we just want to find a price to center the CMM brackets around. One thing that might be worthwhile is to estimate *both* ask and bid prices and take the average. This might produce a slightly better "center" price.

Reply:
Yes, makes sense.
I'll make it so it only caches itself, not the opposite.

To be honest, I was checking and the logic to fetch the price from inch API is currently fetching the inverted price and converting it.
I'll elaborate in a moment

Reply:
According to [1inch api](https://1inch.exchange/#/api), `fromToken` == `baseToken`; `toToken` === `quoteToken`.

So, when I go to the regular 1inch app page and say I want to swap from 1 WETH to USDC, I would expect the equivalent API query to be:

`https://api.1inch.exchange/v1.1/quote?fromTokenSymbol=WETH&toTokenSymbol=USDC&amount=1000000000000000000`

Where amount is `1` unit of base token in wei (note USDC has 6 decimals vs 18 on WETH)

The response returns `fromTokenAmount` === `baseTokenAmount` and `toTokenAmount` === `quoteTokenAmount`

Getting the price out of the response would be then be:

```
price = (quoteTokenAmount/10^quoteTokenDecimals) / (baseTokenAmount/10^baseTokenDecimals) => quoteTokenAmount / baseTokenAmount * 10 ^ - (quoteTokenDecimals - baseTokenDecimals)
```

(Please double check my math, not confident on that)


---

But the price estimation queries the inverted quote/base tokens: https://github.com/gnosis/dex-liquidity-provision/blob/master/scripts/utils/price_utils.js#L134

## Let's try it out both ways:

1. The way it makes more sense to me:
https://api.1inch.exchange/v1.1/quote?fromTokenSymbol=WETH&toTokenSymbol=USDC&amount=1000000000000000000

``` json
{
  "fromToken": {
    "symbol": "WETH",
    "name": "Wrapped Ether",
    "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
    "decimals": 18
  },
  "toToken": {
    "symbol": "USDC",
    "name": "USD Coin",
    "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "decimals": 6
  },
  "toTokenAmount": "352820000",
  "fromTokenAmount": "1000000000000000000",
}
```

Calculating the price:


```
quoteTokenAmount === toTokenAmount = 352820000
baseTokenAmount === fromTokenAmount = 1000000000000000000

price = quoteTokenAmount / baseTokenAmount * 10 ^ - (quoteTokenDecimals/baseTokenDecimals)
price = 352820000 / 1000000000000000000 * 10 ^ - (6 - 18)
price = 352.82
```

2. The way it's implemented on dex-liquidity

https://api.1inch.exchange/v1.1/quote?fromTokenSymbol=USDC&toTokenSymbol=WETH&amount=1000000

```json
{
  "fromToken": {
    "symbol": "USDC",
    "name": "USD Coin",
    "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    "decimals": 6
  },
  "toToken": {
    "symbol": "WETH",
    "name": "Wrapped Ether",
    "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
    "decimals": 18
  },
  "toTokenAmount": "2896822140323620",
  "fromTokenAmount": "1000000",
}
```

Calculating the price:


```
quoteTokenAmount === fromTokenAmount = 1000000
baseTokenAmount === toTokenAmount = 2896822140323620

price = quoteTokenAmount / baseTokenAmount * 10 ^ (baseTokenDecimals/quoteTokenDecimals)
price = 1000000 / 2896822140323620 * 10 ^ (18 - 6)
price = 345.20586751946206
```

---

I'm sure there's a clever math making the results similar, but as I write this I get more confused.

What's your opinion?
I'll hold off calling in Ben as he had enough of price discussions for now =P

Reply:
> According to 1inch api, `fromToken == baseToken`; `toToken == quoteToken`

In both of your examples, `baseToken == WETH` which is the token being priced and `quoteToken == USDC` which is the token the price is in. The base/quote token distinction is only useful for determining the unit of the price and defining trade directions (i.e. ask vs bid trades). Specifically:

1. In this example you are getting a price estimate for selling WETH (`fromToken`) and buying USDC (`toToken`). This is an "ask" order (selling base and buying quote).

2. In this example you are getting a price estimate for buying WETH and selling USDC. This would be a "bid" order in the WETH-USDC market.

So ultimately, which price you decide to use greatly depends on what you are trying to price:
1. Are you trying to estimate a sell price for the base token expressed in quote token? Then you want estimate an ask order price, that is `fromToken == baseToken` and `toToken == quoteToken`.
2. Are you trying to estimate a buy price for the base token expressed in quote token? Then you want to estimate a bid order price, that is `fromToken == quoteToken` and `toToken == baseToken`.

---

That being said, something is very confusing from the numbers you provided. From the prices, you would be able to sell 1 WETH and get 352.00 USDC and then buy it back for only 345.00. This would give you a free 7 USDC!

I think this issue arises from the different sizes of the estimates. I tried the API calls with larger numbers using roughly the same amounts in both price estimates, and I get more expected results:

1. Selling 10.000,00 WETH: https://api.1inch.exchange/v1.1/quote?fromTokenSymbol=WETH&toTokenSymbol=USDC&amount=10000000000000000000000
```js
{
  fromToken: {
    symbol: "WETH",
    // ...
  },
  toToken:{
    symbol: "USDC",
    // ...
  },
  toTokenAmount: "3472087206838",
  fromTokenAmount: "10000000000000000000000",
}
```

This means I would get around 347,21 USDC for each WETH I sold (3.472.087,21 USDC total).

2. Selling 3.500.000,00 USDC: https://api.1inch.exchange/v1.1/quote?fromTokenSymbol=USDC&toTokenSymbol=WETH&amount=3500000000000
```js{
  fromToken: {
    symbol: "USDC",
    // ...
  },
  toToken:{
    symbol: "WETH",
    // ...
  },
  toTokenAmount: "3500000000000",
  fromTokenAmount: "9853439801911923014605",
}
```

This means I would be able to buy each WETH for around 355.21 USDC (9.853,44 WETH total).

Now, the prices make a bit more sense in that if I sell my WETH for USDC and then buy it back, I would be at a loss (which is expected because of slippage and fees).

I hope this helps!

Reply:
Prices make my head go :exploding_head:

Reply:
Ahhhhhhh, so that's the difference! ask vs bid pricing :thinking: 

Thanks a lot for helping me understand.

I'll bring this up with the team to check what is we want to show for these prices.